### PR TITLE
feat: ConfigureProvider to return diagnostics, add diag.USER and UnsquashDiag

### PR DESCRIPTION
**Pull request by disq**
Error field is left for backwards compatibility in proto, it's generated from `response.Diagnostics.Error()`
`diag.USER` is finally there for user-caused errors like bad configuration etc.

### CODE REVIEW

File: provider/provider.go [hunk 1]
```diff
@@ -96,12 +96,16 @@ func (p *Provider) GetProviderConfig(_ context.Context, _ *cqproto.GetProviderCo
 
 func (p *Provider) ConfigureProvider(_ context.Context, request *cqproto.ConfigureProviderRequest) (*cqproto.ConfigureProviderResponse, error) {
```
Comment (disq) [inline provider/provider.go:97]: `error` is unused in the signature but removing it fails the CQServer interface check (which is unrelated I think...)

File: provider/provider.go [hunk 2]
```diff
@@ -96,12 +96,16 @@ func (p *Provider) GetProviderConfig(_ context.Context, _ *cqproto.GetProviderCo
...
 	if p.Logger == nil {
-		return &cqproto.ConfigureProviderResponse{Error: fmt.Sprintf("provider %s logger not defined, make sure to run it with serve", p.Name)}, fmt.Errorf("provider %s logger not defined, make sure to run it with serve", p.Name)
+		return &cqproto.ConfigureProviderResponse{
+			Diagnostics: diag.FromError(fmt.Errorf("provider %s logger not defined, make sure to run it with serve", p.Name), diag.USER),
```
Comment (roneli) [inline provider/provider.go:100]: don't this is a user related error, but more internal as in the provider dev

**Review by roneli — APPROVED** [review]
Had one nit.

File: provider/provider.go (hunk 2 above)
Comment (disq) [reply provider/provider.go]: fixed

File: provider/diag/error.go [hunk 3]
```diff
@@ -201,3 +201,39 @@ func FromError(err error, dt DiagnosticType, opts ...BaseErrorOption) Diagnostic
 		return Diagnostics{NewBaseError(err, dt, opts...)}
 	}
 }
+
+// Mutate generates new Diagnostics (using BaseError) from the given Diagnostics, and applies the given options to each one
+func Mutate(dd Diagnostics, opts ...BaseErrorOption) Diagnostics {
```
Comment (roneli) [inline provider/diag/error.go:206]: where is this used? its kinda cool but still

**Review by roneli — APPROVED** [review]
Looks good, had a question on the mutate diags method

File: provider/diag/error.go (hunk 3 above)
Comment (disq) [reply provider/diag/error.go:206]: @roneli To distinguish between Configure diags vs. Fetch diags. See https://github.com/cloudquery/cloudquery/pull/594 and the last commit in it

Comment (roneli) [reply provider/diag/error.go:206]: I don't think you need to set a resource name / id for diagnostics as its only for "resource diags", but I see the usage

File: provider/diag/diagnostic.go [hunk 4]
```diff
@@ -25,6 +25,8 @@ func (d DiagnosticType) String() string {
 		return "Throttle"
 	case DATABASE:
 		return "Database"
+	case USER:
```
Comment (roneli) [inline provider/diag/diagnostic.go:28]: I know this isn't part of the PR review, but case Internal is missing.

**Review by roneli — CHANGES_REQUESTED** [review]
Added nit

File: provider/diag/diagnostic.go (hunk 4 above)
Comment (disq) [reply provider/diag/diagnostic.go:28]: fixed

File: provider/diag/error.go (hunk 3 above)
Comment (roneli) [reply provider/diag/error.go:206]: So we don't need this function basically, we can remove if not.

File: provider/provider.go (hunk 1 above)
Comment (roneli) [reply provider/provider.go:97]: We should usually avoid returning error and put it in the response, as that causes the GRPC fatal which no user understands why its "fatal"

**Review by roneli — APPROVED** [review]
LGTM, added 2 comments

**Pull request merged**